core/App/ehsm_core_test.cpp

```diff
@@ -42,577 +42,60 @@
-    if (ret != EH_OK) {
-       printf("Failed(%d) to get size of GenerateDataKey!\n", ret);
-       goto cleanup;
-    }
-    printf("GenerateDataKey SUCCESSFULLY!\n");
-    dump_data(plaint_datakey.data, plaint_datakey.datalen);
-    dump_data(cipher_datakey.data, cipher_datakey.datalen);
-
-    /* try to use the cmk to decrypt the datakey */
-    plaintext1.datalen = 0;
-    ret = Decrypt(&cmk, &cipher_datakey, &aad, &plaintext1);
-    if (ret != EH_OK) {
-       printf("Failed(%d) to get size of Decrypt!\n", ret);
-       goto cleanup;
-    }
-    plaintext1.data = (uint8_t*)malloc(plaintext1.datalen);
-    if (plaintext1.data == NULL) {
-       ret = EH_DEVICE_MEMORY;
-       goto cleanup;
-    }
-    ret = Decrypt(&cmk, &cipher_datakey, &aad, &plaintext1);
-    if (ret != EH_OK) {
-       printf("Failed(%d) to Decrypt the datakey!\n", ret);
-       goto cleanup;
-    }
-    printf("Decrypt datakey SUCCESSFULLY!\n");
-    dump_data(plaintext1.data, plaintext1.datalen);
-
-    /* generate a 48 bytes random data key and without plaint text returned */
-    plaint_datakey2.datalen = 48;
-    plaint_datakey2.data = NULL;
-    cipher_datakey2.datalen = 0;
-    ret = GenerateDataKeyWithoutPlaintext(&cmk, &aad, &plaint_datakey2, &cipher_datakey2);
-    if (ret != EH_OK) {
-       printf("Failed(%d) to get size of GenerateDataKeyWithoutPlaintext!\n", ret);
-       goto cleanup;
-    }
-    cipher_datakey2.data = (uint8_t*)malloc(cipher_datakey2.datalen);
-    if (cipher_datakey2.data == NULL) {
-       ret = EH_DEVICE_MEMORY;
-       goto cleanup;
-    }
-
-    ret = GenerateDataKeyWithoutPlaintext(&cmk, &aad, &plaint_datakey2, &cipher_datakey2);
-    if (ret != EH_OK) {
-       printf("Failed(%d) to get size of GenerateDataKey!\n", ret);
-       goto cleanup;
-    }
-    printf("GenerateDataKeyWithoutPlaintext SUCCESSFULLY!\n");
-    dump_data(cipher_datakey2.data, cipher_datakey2.datalen);
-
-    /* try to use the cmk to decrypt the datakey2 */
-    plaintext2.datalen = 0;
-    ret = Decrypt(&cmk, &cipher_datakey2, &aad, &plaintext2);
-    if (ret != EH_OK) {
-       printf("Failed(%d) to get size of Decrypt!\n", ret);
-       goto cleanup;
-    }
-    plaintext2.data = (uint8_t*)malloc(plaintext2.datalen);
-    if (plaintext2.data == NULL) {
-       ret = EH_DEVICE_MEMORY;
-       goto cleanup;
-    }
-    ret = Decrypt(&cmk, &cipher_datakey2, &aad, &plaintext2);
-    if (ret != EH_OK) {
-       printf("Failed(%d) to Decrypt the datakey!\n", ret);
-       goto cleanup;
-    }
-    printf("Decrypt datakey SUCCESSFULLY!\n");
-    dump_data(plaintext2.data, plaintext2.datalen);
-
-cleanup:
-    SAFE_FREE(cmk.keyblob);
-    SAFE_FREE(plaint_datakey.data);
-    SAFE_FREE(cipher_datakey.data);
-    SAFE_FREE(cipher_datakey2.data);
-    SAFE_FREE(plaintext1.data);
-    SAFE_FREE(plaintext2.data);
-
-    printf("============testGenerateDataKey done==========\n");
-    return ret;
+    
+out:
+    printf("============test_AES128 end==========\n");
```

Comment:
seems that you didn't free the memory!！

the NAPI_xxx function will return a json string, i'm not sure whether the nodejs side will free them automatically? and for the c++ unittest, it seems that you need to free them by yourself.

Reply:
It will be fix